[feature]: (#428) validate duplicate voucher code changes

Rakeshsingu opened the pull request:
**📌 What This PR Does**

**This PR adds validation to prevent the reuse of voucher codes across specific submitted transactions.**

✅ Validations Applied To:

    Project

    Sales Order

    Sales Invoice

🔍 What It Checks:

    If a submitted record with the same voucher code and campaign already exists.

    Optionally excludes Repair Orders based on the doctype.

⚠️ Behavior:

    If a duplicate is found, a validation error is thrown with a link to the existing document.

🔄 Why This Is Needed

To ensure that voucher codes are not reused inappropriately across documents, particularly when the associated campaign has voucher validation enabled.

Closes https://github.com/ParaLogicTech/erpnext/issues/428

Rakeshsingu committed:
[feature]: (#428) validate duplicate voucher code changes

Rakeshsingu committed:
[feature]: (#428) clean up changes

Rakeshsingu committed:
[feature]: (#428) clean up

Rakeshsingu committed:
[feature]: (#428) clean up on variable names

Rakeshsingu commented on erpnext/overrides/campaign/campaign_hooks.py:
```diff
@@ -36,3 +51,46 @@ def override_campaign_dashboard(data):
+			additional_filters["is_return"] = 0
+
+		if doctype == "Project":
+			additional_filters["docstatus"] = ["!=", 2]
```
> Validation error if (campaign, campaign_voucher_code) is already used in any non-cancelled Repair Order

As mentioned in the issue description, I have added the docstatus. Let me know if you'd like me to remove it and refactor the code.

SaiFi0102 replied:
no need to check for docstatus for Project doctype

Rakeshsingu commented on erpnext/overrides/campaign/campaign_hooks.py:
```diff
@@ -28,6 +29,20 @@ def validate_campaign_voucher_code(doc):
 				frappe.bold(doc.campaign),
 			))
```
If the voucher code does not pass the regex validation, we throw an error and stop further processing. I believe this should be the first step in the validation flow. Once it passes, we can then proceed to check for duplicates in the system.

@SaiFi0102 Please let me know if you need any further changes or improvements regarding this.

SaiFi0102 replied:
correct and agreed, your code has no major issues here, my suggestion is that instead of breaking the function, it makes more sense to nest it inside if block

SaiFi0102 commented on erpnext/overrides/campaign/campaign_hooks.py:
```diff
@@ -28,6 +29,20 @@ def validate_campaign_voucher_code(doc):
+	if not validate_duplicate_code:
+		return
+
+	project = getattr(doc, "project", None)
```
use doc.get("project") instead

Rakeshsingu committed:
[feature]: (#428) PR changes

SaiFi0102 committed:
Merge remote-tracking branch 'upstream/version-15' into fork/Rakeshsingu/validate_duplicate_voucher_code

SaiFi0102 committed:
fix: PR Review Changes
- improved error message
- improved function naming and params
- fixed non-project cases
- fixed same project document cases
- reordered field position
